### PR TITLE
[consensus][reconfig] enforcing reconfiguration suffix blocks must be empty rule

### PR DESCRIPTION
**Pull request by zekun000**
## Motivation

As our reconfiguration design, any child of reconfiguration block needs to be empty and not executed at all.

We flag the reconfiguration by the executed_state and roll over this flag for any child blocks.

This PR enforces this in two places
- block_store will reject non-empty blocks as reconfiguration child and roll over compute result for empty blocks.
- proposal generator will only generate empty blocks

### Have you read the [Contributing Guidelines on pull requests](https://github.com/libra/libra/blob/master/CONTRIBUTING.md#pull-requests)?

Y
## Test Plan

Added unit tests.

## Related PRs

(If this PR adds or changes functionality, please take some time to update the docs at https://github.com/libra/website, and link to your PR here.)

### CODE REVIEW

**Comment by ShiYour** [comment]
？


---Original---

File: consensus/src/chained_bft/block_storage/block_store.rs [hunk 1]
```diff
@@ -209,21 +218,30 @@ impl<T: Payload> BlockStore<T> {
...
-            .save_tree(vec![block.clone()], vec![])
-            .with_context(|e| format!("Insert block failed with {:?} when saving block", e))?;
-        let new_block = ExecutedBlock::new(block, compute_res);
-        self.inner.write().unwrap().insert_block(new_block)
+        // Reconfiguration rule - if a block is a child of reconfiguration, it needs to be empty
+        // So we roll over the executed state until it's committed and we start new epoch.
+        let parent_state = parent_block.compute_result();
+        let compute_res = if parent_state.executed_state.validators.is_some() {
+            ensure!(
+                block.payload().filter(|p| **p != T::default()).is_none(),
```
Comment (dmitri-perelman) [inline consensus/src/chained_bft/block_storage/block_store.rs]: Why is it `**p`?

File: consensus/src/chained_bft/block_storage/block_store.rs [hunk 2]
```diff
@@ -209,21 +218,30 @@ impl<T: Payload> BlockStore<T> {
...
+                "Reconfiguration suffix should not carry payload"
+            );
+            StateComputeResult {
+                executed_state: parent_state.executed_state.clone(),
+                compute_status: vec![],
+            }
+        } else {
+            // We still execute NIL block for the block prologue transaction.
```
Comment (dmitri-perelman) [inline consensus/src/chained_bft/block_storage/block_store.rs]: Hmm I don't quite understand this comment: why do we talk about NIL blocks here?

File: consensus/src/chained_bft/liveness/proposal_generator.rs [hunk 3]
```diff
@@ -199,19 +200,26 @@ impl<T: Payload> ProposalGenerator<T> {
...
-        let block_store = Arc::clone(&self.block_store);
-        match self
-            .txn_manager
-            .pull_txns(self.max_block_size, exclude_payload)
-            .await
+        // Reconfiguration rule - we propose empty blocks after reconfiguration until it's committed
+        let txns = if hqc_block
+            .compute_result()
+            .executed_state
+            .validators
```
Comment (dmitri-perelman) [inline consensus/src/chained_bft/liveness/proposal_generator.rs]: nit: may be add a wrapper function `is_reconfiguration()` to the compute result?

File: consensus/src/chained_bft/block_storage/block_store.rs (hunk 1 above)
Comment (zekun000) [reply consensus/src/chained_bft/block_storage/block_store.rs]: well, payload() returns a Option<&T> and filter takes a P: FnOnce(&T) for Option<T> so p becomes &&T

File: consensus/src/chained_bft/block_storage/block_store.rs (hunk 2 above)
Comment (zekun000) [reply consensus/src/chained_bft/block_storage/block_store.rs]: let me rephrase it, NIL blocks don't have payload at all, but we still pass a T::default() to executor because we'll inject a special transaction.
